tests/e2e/attention/generate_e2e_attention_tests.py

```diff
@@ -0,0 +1,487 @@
+    m = shape.m
+    k1 = shape.k1
+    k2 = shape.k2
+    n = shape.n
+    scale = shape.scale
+
+    shapes_scale = TestInputTensorShapes(
+        batch=batch,
+        m=m,
+        k1=k1,
+        k2=k2,
+        n=n,
+        scale=scale,
+    )
+    return shapes_scale
+
+
+# Helper to return input, kernel and output shapes based on the layout and the Attention Params.
+def get_tensor_shapes(
+    shapes_scale: TestShapeAndScale,
+):
+    batch = shapes_scale.batch
+    m = shapes_scale.m
+    k1 = shapes_scale.k1
+    k2 = shapes_scale.k2
+    n = shapes_scale.n
+    scale = shapes_scale.scale
+
+    query_tensor_shape = [batch, m, k1]
+    key_tensor_shape = [batch, k2, k1]
+    value_tensor_shape = [batch, k2, n]
+    result_tensor_shape = [batch, m, n]
+
+    return query_tensor_shape, key_tensor_shape, value_tensor_shape, result_tensor_shape
+
+
+# Helper for generate_function.
+# Generates a name for a test function in the generated MLIR code.
+def generate_function_name(
+    query_type: QueryElemTypeId,
+    key_type: KeyElemTypeId,
+    value_type: ValueElemTypeId,
+    shapes_scale: TestInputTensorShapes,
+):
+    query_t = query_type.value
+    key_t = key_type.value
+    value_t = value_type.value
+    result_t = value_type.value
+
+    batch = shapes_scale.batch
+    m = shapes_scale.m
+    k1 = shapes_scale.k1
+    k2 = shapes_scale.k2
+    n = shapes_scale.n
+
+    attention = "attention"
+    return (
+        f"{attention}_{batch}_{m}_{k1}_{k2}_{n}"
+        + f"_dtype_{query_t}_{key_t}_{value_t}_{result_t}"
+    )
+
+
+# Represents a generated test function.
+@dataclasses.dataclass
+class MLIRFunction:
+    name: str
+    signature: str
+    import_declaration: str
+    definition: str
+
+
+# Generates a test function in the generated MLIR code.
+# The generated function will take the same arguments as iree_linalg_ext.attention variants
+# and will just call iree_linalg_ext.attention variants with them, returning its result.
+def generate_function(
+    query_type: QueryElemTypeId,
+    key_type: KeyElemTypeId,
+    value_type: ValueElemTypeId,
+    shape_scale: TestShapeAndScale,
+):
+    shapes_scale = generate_shapes_and_scale(shape_scale)
+    func_name = generate_function_name(
+        query_type,
+        key_type,
+        value_type,
+        shapes_scale,
+    )
+
+    query_shape, key_shape, value_shape, result_shape = get_tensor_shapes(shapes_scale)
+    query_tensor_type = (
+        f"tensor<{query_shape[0]}x{query_shape[1]}x{query_shape[2]}x{query_type.value}>"
+    )
+    key_tensor_type = (
+        f"tensor<{key_shape[0]}x{key_shape[1]}x{key_shape[2]}x{key_type.value}>"
+    )
+    value_tensor_type = (
+        f"tensor<{value_shape[0]}x{value_shape[1]}x{value_shape[2]}x{value_type.value}>"
+    )
+    result_tensor_type = f"tensor<{result_shape[0]}x{result_shape[1]}x{result_shape[2]}x{value_type.value}>"
+    F32 = "f32"
+    F16 = "f16"
+    op_name = "iree_linalg_ext.attention"
```

Comment:
How stable is the op definition of `iree_linalg_ext.attention`? I'd like to focus testing on what frontends are producing, but this op is defined directly in IREE, unlike upstream `linalg`, `torch_mlir`, `torch`, `stablehlo`, etc. dialect ops.

I asked around on Discord for more info here: https://discord.com/channels/689900678990135345/908461811026767952/1272640489924857908

I'm concerned that if we move these tests out of tree, it will be harder to rely on an IREE op than, say, `torch.nn.functional.scaled_dot_product_attention` and letting the frontend handle lowering to IREE ops. That being said, it looks like in https://github.com/nod-ai/sharktank (our frontend for LLMs / gen-ai models) we already directly emit `iree_linalg_ext.attention` in some cases.

It looked like Rohan was changing the op definition based on discussion at https://discord.com/channels/689900678990135345/689906000043573354/1271261222280364042 . That sort of change will need some effort to incorporate into this string-based generator.

Reply:
@Groverkss @raikonenfnu @pashu123 would have more context than me.

Reply:
The op definition of `iree_linalg_ext.attention` is for now, stable. Only additions we will be doing for now are adding optional entries to the operation, which shouldn't change what the string based emitter emits. We plan to later move to a different op definition.

I would prefer if we tested `iree_linalg_ext.attention` directly. Using `torch.nn.functional.scaled_dot_product_attention` ties us to what that operation limits us to, which we cannot control. Ideally, in future, we want to be emitting `iree_linalg_ext.online_attention` + elementwise operations, and test it directly. There is no operation in torch that can represent `online_attention` in any way.